diff --git a/Makefile b/Makefile
index 39a99d7..6978d95 100644
--- a/Makefile
+++ b/Makefile
@@ -132,6 +132,8 @@ UPROGS=\
 	$U/_grind\
 	$U/_wc\
 	$U/_zombie\
+	$U/_dummyproc\
+	$U/_testprocinfo\
 
 fs.img: mkfs/mkfs README $(UPROGS)
 	mkfs/mkfs fs.img README $(UPROGS)
@@ -153,7 +155,7 @@ QEMUGDB = $(shell if $(QEMU) -help | grep -q '^-gdb'; \
 	then echo "-gdb tcp::$(GDBPORT)"; \
 	else echo "-s -p $(GDBPORT)"; fi)
 ifndef CPUS
-CPUS := 3
+CPUS := 1
 endif
 
 QEMUOPTS = -machine virt -bios none -kernel $K/kernel -m 128M -smp $(CPUS) -nographic
diff --git a/kernel/defs.h b/kernel/defs.h
index a3c962b..7c7ec16 100644
--- a/kernel/defs.h
+++ b/kernel/defs.h
@@ -106,6 +106,9 @@ void            yield(void);
 int             either_copyout(int user_dst, uint64 dst, void *src, uint64 len);
 int             either_copyin(void *dst, int user_src, uint64 src, uint64 len);
 void            procdump(void);
+int             getpinfo(uint64); // offline 3
+void            updateAliveProc();
+void            priority_boosting();
 
 // swtch.S
 void            swtch(struct context*, struct context*);
diff --git a/kernel/param.h b/kernel/param.h
index 6624bff..75ac8c2 100644
--- a/kernel/param.h
+++ b/kernel/param.h
@@ -11,3 +11,8 @@
 #define NBUF         (MAXOPBLOCKS*3)  // size of disk block cache
 #define FSSIZE       2000  // size of file system in blocks
 #define MAXPATH      128   // maximum file path name
+//Offline 3
+#define TIME_LIMIT_1 1
+#define TIME_LIMIT_2 2
+#define BOOST_INTERVAL 64
+#define DEFAULT_TICKET_COUNT 10
diff --git a/kernel/proc.c b/kernel/proc.c
index 959b778..70e480e 100644
--- a/kernel/proc.c
+++ b/kernel/proc.c
@@ -5,6 +5,8 @@
 #include "spinlock.h"
 #include "proc.h"
 #include "defs.h"
+#include "pstat.h"
+#include "rand.h"
 
 struct cpu cpus[NCPU];
 
@@ -26,6 +28,23 @@ extern char trampoline[]; // trampoline.S
 // must be acquired before any p->lock.
 struct spinlock wait_lock;
 
+//Offline 3
+int getRunnableProcTickets(void)
+{
+  struct proc* p;
+  int total = 0;
+
+  for(p = proc; p < &proc[NPROC]; p++)
+  {
+    if(p->state == RUNNABLE){
+      acquire(&p->lock);
+      total += p->ticket_current;  
+      release(&p->lock);
+    }
+  }
+  return total;
+}
+
 // Allocate a page for each process's kernel stack.
 // Map it high in memory, followed by an invalid
 // guard page.
@@ -114,6 +133,11 @@ allocproc(void)
   for(p = proc; p < &proc[NPROC]; p++) {
     acquire(&p->lock);
     if(p->state == UNUSED) {
+      p->ticket_original = DEFAULT_TICKET_COUNT;
+      p->ticket_current = DEFAULT_TICKET_COUNT;
+      p->time_slice = 0;
+      p->queue = 1;
+      p->consumed_time = 0;
       goto found;
     } else {
       release(&p->lock);
@@ -319,6 +343,11 @@ fork(void)
   release(&wait_lock);
 
   acquire(&np->lock);
+  np->ticket_original = p->ticket_original; 
+  np->ticket_current = p->ticket_original;
+  np->time_slice = 0;
+  np->queue = 1;
+  np->consumed_time = 0;
   np->state = RUNNABLE;
   release(&np->lock);
 
@@ -446,28 +475,88 @@ scheduler(void)
 {
   struct proc *p;
   struct cpu *c = mycpu();
-  
+  int queue_number = 1;  // Current queue number (start with the topmost queue)
   c->proc = 0;
   for(;;){
     // Avoid deadlock by ensuring that devices can interrupt.
     intr_on();
 
+    int runnableTickets = getRunnableProcTickets();
+    if( runnableTickets == 0 ){
+      for(p = proc; p < &proc[NPROC]; p++) {
+        acquire(&p->lock);
+        p->ticket_current = p->ticket_original;
+        release(&p->lock);
+      }
+    }
+    int randInt = genrand(runnableTickets)+1;
+    int totalTicket = 0;
+
     for(p = proc; p < &proc[NPROC]; p++) {
       acquire(&p->lock);
-      if(p->state == RUNNABLE) {
-        // Switch to chosen process.  It is the process's job
-        // to release its lock and then reacquire it
-        // before jumping back to us.
-        p->state = RUNNING;
-        c->proc = p;
-        swtch(&c->context, &p->context);
-
-        // Process is done running for now.
-        // It should have changed its p->state before coming back.
-        c->proc = 0;
+      if(p->state == RUNNABLE && p->queue == queue_number) {
+        if(queue_number==1){
+          totalTicket+=p->ticket_current;
+          if( totalTicket >= randInt && p->ticket_current != 0 ){
+            // Switch to chosen process.  It is the process's job
+            // to release its lock and then reacquire it
+            // before jumping back to us.
+            p->state = RUNNING;
+            c->proc = p;
+            p->ticket_current-=1;
+            p->consumed_time = 0;
+            swtch(&c->context, &p->context);
+            
+
+            // Process is done running for now.
+            // It should have changed its p->state before coming back.
+            c->proc = 0;
+
+            if (p->state == RUNNABLE){
+              if(p->consumed_time>=TIME_LIMIT_1){
+                p->queue = 2;
+              }
+            }
+          }
+        }
+        else{
+          // Switch to chosen process.  It is the process's job
+          // to release its lock and then reacquire it
+          // before jumping back to us.
+          p->state = RUNNING;
+          c->proc = p;
+          p->consumed_time = 0;
+          swtch(&c->context, &p->context);
+
+          // Process is done running for now.
+          // It should have changed its p->state before coming back.
+          c->proc = 0;
+
+          if(p->consumed_time<TIME_LIMIT_2){
+            p->queue = 1;
+          }
+        }
+        // Reset the queue number to start from the topmost queue
+        queue_number = 1;
+        release(&p->lock);
+        break;
       }
       release(&p->lock);
     }
+    // Move to the next queue if no process was found in the current queue
+    if (p >= &proc[NPROC]) {
+      queue_number = (queue_number == 1) ? 2 : 1;
+    }
+    
+  }
+}
+
+void priority_boosting(){
+  struct proc *p;
+  for (p = proc; p < &proc[NPROC]; p++) {
+    acquire(&p->lock);
+    p->queue = 1;
+    release(&p->lock);
   }
 }
 
@@ -681,3 +770,36 @@ procdump(void)
     printf("\n");
   }
 }
+
+int getpinfo(uint64 addr){
+  struct proc *curProc = myproc();
+  struct proc *p;
+  struct pstat ps;
+  int i=0;
+  for(p = proc; p < &proc[NPROC]; p++){
+    acquire(&p->lock);
+    ps.pid[i] = p->pid;
+    ps.inuse[i] = p->state != UNUSED;
+    ps.inQ[i] = p->queue;
+    ps.tickets_original[i] = p->ticket_original;
+    ps.tickets_current[i] = p->ticket_current;
+    ps.time_slices[i] = p->time_slice;
+    release(&p->lock);
+    i++;
+  }
+  if(copyout(curProc->pagetable, addr, (char *)&ps, sizeof(ps)) < 0)
+    return -1;
+  return 0;
+}
+
+void updateAliveProc(){
+  struct proc *p;
+  for (p = proc; p < &proc[NPROC]; p++) {
+    acquire(&p->lock);
+    if (p->state == RUNNING) {
+      p->time_slice += 1;
+      p->consumed_time+=1;
+    }
+    release(&p->lock);
+  }
+}
diff --git a/kernel/proc.h b/kernel/proc.h
index d021857..7379719 100644
--- a/kernel/proc.h
+++ b/kernel/proc.h
@@ -104,4 +104,11 @@ struct proc {
   struct file *ofile[NOFILE];  // Open files
   struct inode *cwd;           // Current directory
   char name[16];               // Process name (debugging)
+
+  // offline 3
+  int queue;                   // Queue number for MLFQ
+  int ticket_original; 
+  int ticket_current; 
+  int time_slice;   
+  int consumed_time;
 };
diff --git a/kernel/pstat.h b/kernel/pstat.h
new file mode 100644
index 0000000..60f6a3f
--- /dev/null
+++ b/kernel/pstat.h
@@ -0,0 +1,12 @@
+#ifndef _PSTAT_H_
+#define _PSTAT_H_
+#include "param.h"
+struct pstat {
+int pid[NPROC]; // the process ID of each process
+int inuse[NPROC]; // whether this slot of the process table is being used (1 or 0)
+int inQ[NPROC]; // which queue the process is currently in
+int tickets_original[NPROC]; // the number of tickets each process originally had
+int tickets_current[NPROC]; // the number of tickets each process currently has
+int time_slices[NPROC]; // the number of time slices each process has been scheduled
+};
+#endif // _PSTAT_H_
diff --git a/kernel/rand.h b/kernel/rand.h
new file mode 100644
index 0000000..802ab1e
--- /dev/null
+++ b/kernel/rand.h
@@ -0,0 +1,114 @@
+/* A C-program for MT19937: Real number version  (1998/4/6)    */
+/*   genrand() generates one pseudorandom real number (double) */
+/* which is uniformly distributed on [0,1]-interval, for each  */
+/* call. sgenrand(seed) set initial values to the working area */
+/* of 624 words. Before genrand(), sgenrand(seed) must be      */
+/* called once. (seed is any 32-bit integer except for 0).     */
+/* Integer generator is obtained by modifying two lines.       */
+/*   Coded by Takuji Nishimura, considering the suggestions by */
+/* Topher Cooper and Marc Rieffel in July-Aug. 1997.           */
+
+/* This library is free software; you can redistribute it and/or   */
+/* modify it under the terms of the GNU Library General Public     */
+/* License as published by the Free Software Foundation; either    */
+/* version 2 of the License, or (at your option) any later         */
+/* version.                                                        */
+/* This library is distributed in the hope that it will be useful, */
+/* but WITHOUT ANY WARRANTY; without even the implied warranty of  */
+/* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.            */
+/* See the GNU Library General Public License for more details.    */
+/* You should have received a copy of the GNU Library General      */
+/* Public License along with this library; if not, write to the    */
+/* Free Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA   */ 
+/* 02111-1307  USA                                                 */
+
+/* Copyright (C) 1997 Makoto Matsumoto and Takuji Nishimura.       */
+/* When you use this, send an email to: matumoto@math.keio.ac.jp   */
+/* with an appropriate reference to your work.                     */
+
+/* REFERENCE                                                       */
+/* M. Matsumoto and T. Nishimura,                                  */
+/* "Mersenne Twister: A 623-Dimensionally Equidistributed Uniform  */
+/* Pseudo-Random Number Generator",                                */
+/* ACM Transactions on Modeling and Computer Simulation,           */
+/* Vol. 8, No. 1, January 1998, pp 3--30.                          */
+
+// #include<stdio.h>
+
+/* Period parameters */  
+#define N 624
+#define M 397
+#define MATRIX_A 0x9908b0df   /* constant vector a */
+#define UPPER_MASK 0x80000000 /* most significant w-r bits */
+#define LOWER_MASK 0x7fffffff /* least significant r bits */
+
+/* Tempering parameters */   
+#define TEMPERING_MASK_B 0x9d2c5680
+#define TEMPERING_MASK_C 0xefc60000
+#define TEMPERING_SHIFT_U(y)  (y >> 11)
+#define TEMPERING_SHIFT_S(y)  (y << 7)
+#define TEMPERING_SHIFT_T(y)  (y << 15)
+#define TEMPERING_SHIFT_L(y)  (y >> 18)
+#define RAND_MAX 0x7fffffff
+static unsigned long mt[N]; /* the array for the state vector  */
+static int mti=N+1; /* mti==N+1 means mt[N] is not initialized */
+static int set = 0;
+/* initializing the array with a NONZERO seed */
+void
+sgenrand(unsigned long seed)	
+{
+    /* setting initial seeds to mt[N] using         */
+    /* the generator Line 25 of Table 1 in          */
+    /* [KNUTH 1981, The Art of Computer Programming */
+    /*    Vol. 2 (2nd Ed.), pp102]                  */
+    mt[0]= seed & 0xffffffff;
+    for (mti=1; mti<N; mti++)
+        mt[mti] = (69069 * mt[mti-1]) & 0xffffffff;
+}
+
+long  /* for integer generation */
+genrand(int upper)
+{
+    if( set == 0 ){
+        sgenrand(4357);
+        set = 1;
+    }
+    
+    unsigned long y;
+    static unsigned long mag01[2]={0x0, MATRIX_A};
+    /* mag01[x] = x * MATRIX_A  for x=0,1 */
+
+    if (mti >= N) { /* generate N words at one time */
+        int kk;
+
+        if (mti == N+1)   /* if sgenrand() has not been called, */
+            sgenrand(4357); /* a default initial seed is used   */
+
+        for (kk=0;kk<N-M;kk++) {
+            y = (mt[kk]&UPPER_MASK)|(mt[kk+1]&LOWER_MASK);
+            mt[kk] = mt[kk+M] ^ (y >> 1) ^ mag01[y & 0x1];
+        }
+        for (;kk<N-1;kk++) {
+            y = (mt[kk]&UPPER_MASK)|(mt[kk+1]&LOWER_MASK);
+            mt[kk] = mt[kk+(M-N)] ^ (y >> 1) ^ mag01[y & 0x1];
+        }
+        y = (mt[N-1]&UPPER_MASK)|(mt[0]&LOWER_MASK);
+        mt[N-1] = mt[M-1] ^ (y >> 1) ^ mag01[y & 0x1];
+
+        mti = 0;
+    }
+  
+    y = mt[mti++];
+    y ^= TEMPERING_SHIFT_U(y);
+    y ^= TEMPERING_SHIFT_S(y) & TEMPERING_MASK_B;
+    y ^= TEMPERING_SHIFT_T(y) & TEMPERING_MASK_C;
+    y ^= TEMPERING_SHIFT_L(y);
+   // int mx = 1000;
+   // double val1 = ( (double)y * 2.3283064370807974e-10 );
+   // int val2 = val1*mx;
+   // return val2;
+    //return ( (double)y * 2.3283064370807974e-10 ); /* reals */
+    y = y & RAND_MAX;
+    y = y % upper;
+    return y;  /* for integer generation */
+}
\ No newline at end of file
diff --git a/kernel/syscall.c b/kernel/syscall.c
index ed65409..b913a9c 100644
--- a/kernel/syscall.c
+++ b/kernel/syscall.c
@@ -101,6 +101,9 @@ extern uint64 sys_unlink(void);
 extern uint64 sys_link(void);
 extern uint64 sys_mkdir(void);
 extern uint64 sys_close(void);
+// offline 3
+extern uint64 sys_settickets(void);
+extern uint64 sys_getpinfo(void); 
 
 // An array mapping syscall numbers from syscall.h
 // to the function that handles the system call.
@@ -126,6 +129,8 @@ static uint64 (*syscalls[])(void) = {
 [SYS_link]    sys_link,
 [SYS_mkdir]   sys_mkdir,
 [SYS_close]   sys_close,
+[SYS_settickets]  sys_settickets, 
+[SYS_getpinfo]  sys_getpinfo, 
 };
 
 void
diff --git a/kernel/syscall.h b/kernel/syscall.h
index bc5f356..e749b86 100644
--- a/kernel/syscall.h
+++ b/kernel/syscall.h
@@ -20,3 +20,6 @@
 #define SYS_link   19
 #define SYS_mkdir  20
 #define SYS_close  21
+#define SYS_settickets 22
+#define SYS_getpinfo 23
+
diff --git a/kernel/sysproc.c b/kernel/sysproc.c
index 1de184e..a5e8cf4 100644
--- a/kernel/sysproc.c
+++ b/kernel/sysproc.c
@@ -5,6 +5,7 @@
 #include "memlayout.h"
 #include "spinlock.h"
 #include "proc.h"
+#include "pstat.h"
 
 uint64
 sys_exit(void)
@@ -89,3 +90,28 @@ sys_uptime(void)
   release(&tickslock);
   return xticks;
 }
+
+// offline 3
+uint64
+sys_settickets(void)
+{
+  int tickets;
+  argint(0,&tickets);
+  if( tickets < 0)
+    return -1;
+  
+  struct proc *curProc = myproc();
+  acquire(&curProc->lock);
+  curProc->ticket_original = tickets;
+  curProc->ticket_current = tickets;
+  release(&curProc->lock);
+  return 0;
+}
+
+uint64
+sys_getpinfo(void)
+{
+  uint64 addr;
+  argaddr(0, &addr);
+  return getpinfo(addr); 
+}
diff --git a/kernel/trap.c b/kernel/trap.c
index 512c850..b89485a 100644
--- a/kernel/trap.c
+++ b/kernel/trap.c
@@ -77,8 +77,11 @@ usertrap(void)
     exit(-1);
 
   // give up the CPU if this is a timer interrupt.
-  if(which_dev == 2)
-    yield();
+  if(which_dev == 2){
+    if((myproc()->queue==1 && myproc()->consumed_time>=TIME_LIMIT_1) ||(myproc()->queue==2 && myproc()->consumed_time>=TIME_LIMIT_2))
+      yield();
+
+  }
 
   usertrapret();
 }
@@ -151,8 +154,11 @@ kerneltrap()
   }
 
   // give up the CPU if this is a timer interrupt.
-  if(which_dev == 2 && myproc() != 0 && myproc()->state == RUNNING)
-    yield();
+  if(which_dev == 2 && myproc() != 0 && myproc()->state == RUNNING){
+    if((myproc()->queue==1 && myproc()->consumed_time>=TIME_LIMIT_1) ||(myproc()->queue==2 && myproc()->consumed_time>=TIME_LIMIT_2))
+      yield();
+
+  }
 
   // the yield() may have caused some traps to occur,
   // so restore trap registers for use by kernelvec.S's sepc instruction.
@@ -162,11 +168,19 @@ kerneltrap()
 
 void
 clockintr()
-{
+{ 
   acquire(&tickslock);
   ticks++;
+  int curTick = ticks;
   wakeup(&ticks);
   release(&tickslock);
+  updateAliveProc();
+  if(curTick%BOOST_INTERVAL==0){
+    priority_boosting();
+  }
+    
+  
+  
 }
 
 // check if it's an external interrupt or software interrupt,
diff --git a/user/dummyproc.c b/user/dummyproc.c
new file mode 100644
index 0000000..3f471fb
--- /dev/null
+++ b/user/dummyproc.c
@@ -0,0 +1,23 @@
+#include "kernel/types.h"
+#include "kernel/stat.h"
+#include "user/user.h"
+
+int
+main(int argc, char *argv[])
+{
+  if(argc<2)
+  {
+    fprintf(2, "Usage: %s sys_call_num command\n", argv[0]);
+    exit(1);
+  }
+  int ticket = atoi(argv[1]);
+  if (settickets(ticket) < 0) 
+  {
+    fprintf(2, "%s: exec failed\n", argv[0]);
+    exit(1);   
+  }
+  if(fork()==0){
+    while (1);
+  }
+  exit(0);
+}
\ No newline at end of file
diff --git a/user/testprocinfo.c b/user/testprocinfo.c
new file mode 100644
index 0000000..a5be577
--- /dev/null
+++ b/user/testprocinfo.c
@@ -0,0 +1,22 @@
+#include "kernel/types.h"
+#include "kernel/stat.h"
+#include "user/user.h"
+#include "kernel/pstat.h"
+int
+main(int argc,char *argv[]){
+    struct pstat st;
+    if (getpinfo(&st) < 0) 
+    {
+        fprintf(2, "%s: exec failed\n", argv[0]);
+        exit(1);
+    }
+    printf("PID | In Use | inQ | Original Tickets | Current Tickets | Time Slices\n");
+    for(int i=0;i<NPROC;i++)
+    {
+        if( st.inuse[i] != 0 )
+        {
+          printf("%d\t%d\t%d\t\t%d\t\t%d\t\t%d\n",st.pid[i],st.inuse[i],st.inQ[i],st.tickets_original[i],st.tickets_current[i],st.time_slices[i]);
+        }
+    }
+    exit(0);
+}
\ No newline at end of file
diff --git a/user/user.h b/user/user.h
index 4d398d5..f22877d 100644
--- a/user/user.h
+++ b/user/user.h
@@ -1,4 +1,5 @@
 struct stat;
+struct pstat; //offline 3
 
 // system calls
 int fork(void);
@@ -22,6 +23,9 @@ int getpid(void);
 char* sbrk(int);
 int sleep(int);
 int uptime(void);
+// offline 3
+int settickets(int); 
+int getpinfo(struct pstat*);
 
 // ulib.c
 int stat(const char*, struct stat*);
diff --git a/user/usys.pl b/user/usys.pl
index 01e426e..be025c2 100755
--- a/user/usys.pl
+++ b/user/usys.pl
@@ -36,3 +36,5 @@ entry("getpid");
 entry("sbrk");
 entry("sleep");
 entry("uptime");
+entry("settickets"); 
+entry("getpinfo"); 
\ No newline at end of file
